<?xml version="1.0"?>
<robot xmlns:xacro="http://wiki.ros.org/xacro" name="ur5">

    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
        </plugin>
      </gazebo>
      <link name="base_link">
        <visual>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/visual/base.dae"/>
          </geometry>
          <material name="LightGrey">
            <color rgba="0.7 0.7 0.7 1.0"/>
          </material>
        </visual>
        <collision>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/collision/base.stl"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="4.0"/>
          <!-- <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/> -->
          <origin rpy="0 0 0" xyz="0.0 0.0 0.085"/>
          <inertia ixx="0.00443333156" ixy="0.0" ixz="0.0" iyy="0.00443333156" iyz="0.0" izz="0.0072"/>
        </inertial>
      </link>
      <joint name="shoulder_pan_joint" type="revolute">
        <parent link="base_link"/>
        <child link="shoulder_link"/>
        <!-- <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.089159"/> -->
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.154159"/>
        <axis xyz="0 0 1"/>
        <limit effort="300.0" lower="-3.14159265" upper="3.14159265" velocity="3.15"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>
      <link name="shoulder_link">
        <visual>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/visual/shoulder.dae"/>
          </geometry>
          <material name="LightGrey">
            <color rgba="0.7 0.7 0.7 1.0"/>
          </material>
        </visual>
        <collision>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/collision/shoulder.stl"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="3.7"/>
          <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
          <inertia ixx="0.010267495893" ixy="0.0" ixz="0.0" iyy="0.010267495893" iyz="0.0" izz="0.00666"/>
        </inertial>
      </link>
      <joint name="shoulder_lift_joint" type="revolute">
        <parent link="shoulder_link"/>
        <child link="upper_arm_link"/>
        <origin rpy="0.0 1.570796325 0.0" xyz="0.0 0.13585 0.0"/>
        <axis xyz="0 1 0"/>
        <limit effort="300.0" lower="-3.14159265" upper="3.14159265" velocity="3.15"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>
      <link name="upper_arm_link">
        <visual>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/visual/upperarm.dae"/>
          </geometry>
          <material name="LightGrey">
            <color rgba="0.7 0.7 0.7 1.0"/>
          </material>
        </visual>
        <collision>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/collision/upperarm.stl"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="8.393"/>
          <origin rpy="0 0 0" xyz="0.0 0.0 0.28"/>
          <inertia ixx="0.22689067591" ixy="0.0" ixz="0.0" iyy="0.22689067591" iyz="0.0" izz="0.0151074"/>
        </inertial>
      </link>
      <joint name="elbow_joint" type="revolute">
        <parent link="upper_arm_link"/>
        <child link="forearm_link"/>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 -0.1197 0.425"/>
        <axis xyz="0 1 0"/>
        <limit effort="300.0" lower="-3.14159265" upper="3.14159265" velocity="3.15"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>
      <link name="forearm_link">
        <visual>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/visual/forearm.dae"/>
          </geometry>
          <material name="LightGrey">
            <color rgba="0.7 0.7 0.7 1.0"/>
          </material>
        </visual>
        <collision>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/collision/forearm.stl"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="2.275"/>
          <origin rpy="0 0 0" xyz="0.0 0.0 0.25"/>
          <inertia ixx="0.049443313556" ixy="0.0" ixz="0.0" iyy="0.049443313556" iyz="0.0" izz="0.004095"/>
        </inertial>
      </link>
      <joint name="wrist_1_joint" type="revolute">
        <parent link="forearm_link"/>
        <child link="wrist_1_link"/>
        <origin rpy="0.0 1.570796325 0.0" xyz="0.0 0.0 0.39225"/>
        <axis xyz="0 1 0"/>
        <limit effort="200.0" lower="-3.14159265" upper="3.14159265" velocity="3.2"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>
      <link name="wrist_1_link">
        <visual>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/visual/wrist1.dae"/>
          </geometry>
          <material name="LightGrey">
            <color rgba="0.7 0.7 0.7 1.0"/>
          </material>
        </visual>
        <collision>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/collision/wrist1.stl"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="1.219"/>
          <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
          <inertia ixx="0.111172755531" ixy="0.0" ixz="0.0" iyy="0.111172755531" iyz="0.0" izz="0.21942"/>
        </inertial>
      </link>
      <joint name="wrist_2_joint" type="revolute">
        <parent link="wrist_1_link"/>
        <child link="wrist_2_link"/>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.093 0.0"/>
        <axis xyz="0 0 1"/>
        <limit effort="200.0" lower="-3.14159265" upper="3.14159265" velocity="3.2"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>
      <link name="wrist_2_link">
        <visual>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/visual/wrist2.dae"/>
          </geometry>
          <material name="LightGrey">
            <color rgba="0.7 0.7 0.7 1.0"/>
          </material>
        </visual>
        <collision>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/collision/wrist2.stl"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="1.219"/>
          <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
          <inertia ixx="0.111172755531" ixy="0.0" ixz="0.0" iyy="0.111172755531" iyz="0.0" izz="0.21942"/>
        </inertial>
      </link>
      <joint name="wrist_3_joint" type="revolute">
        <parent link="wrist_2_link"/>
        <child link="wrist_3_link"/>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.09465"/>
        <axis xyz="0 1 0"/>
        <limit effort="200.0" lower="-3.14159265" upper="3.14159265" velocity="3.2"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>
      <link name="wrist_3_link">
        <visual>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/visual/wrist3.dae"/>
          </geometry>
          <material name="LightGrey">
            <color rgba="0.7 0.7 0.7 1.0"/>
          </material>
        </visual>
        <collision>
          <geometry>
            <mesh filename="package://ur_description/meshes/ur5/collision/wrist3.stl"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="0.1879"/>
          <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
          <inertia ixx="0.0171364731454" ixy="0.0" ixz="0.0" iyy="0.0171364731454" iyz="0.0" izz="0.033822"/>
        </inertial>
      </link>
      <joint name="ee_fixed_joint" type="fixed">
        <parent link="wrist_3_link"/>
        <child link="ee_link"/>
        <origin rpy="0.0 0.0 1.570796325" xyz="0.0 0.0823 0.0"/>
      </joint>
      <link name="ee_link">
        <collision>
          <geometry>
            <box size="0.01 0.01 0.01"/>
          </geometry>
          <origin rpy="0 0 0" xyz="-0.01 0 0"/>
        </collision>
      </link>
      <transmission name="shoulder_pan_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="shoulder_pan_joint">
          <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="shoulder_pan_motor">
          <mechanicalReduction>1</mechanicalReduction>
        </actuator>
      </transmission>
      <transmission name="shoulder_lift_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="shoulder_lift_joint">
          <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="shoulder_lift_motor">
          <mechanicalReduction>1</mechanicalReduction>
        </actuator>
      </transmission>
      <transmission name="elbow_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="elbow_joint">
          <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="elbow_motor">
          <mechanicalReduction>1</mechanicalReduction>
        </actuator>
      </transmission>
      <transmission name="wrist_1_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="wrist_1_joint">
          <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wrist_1_motor">
          <mechanicalReduction>1</mechanicalReduction>
        </actuator>
      </transmission>
      <transmission name="wrist_2_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="wrist_2_joint">
          <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wrist_2_motor">
          <mechanicalReduction>1</mechanicalReduction>
        </actuator>
      </transmission>
      <transmission name="wrist_3_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="wrist_3_joint">
          <hardwareInterface>EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wrist_3_motor">
          <mechanicalReduction>1</mechanicalReduction>
        </actuator>
      </transmission>
  
  
  <!-- <xacro:include filename="$(find myur5_description)/urdf/lidar/lidarD1.xacro" /> -->
<joint name="laser_joint_D1" type="fixed">
    <parent link="wrist_3_link"/>
    <child link="laser_frame_D1"/>
    <origin xyz="0.034 0.04 -0.034" rpy="0.7458527 0.2740142 1.2858722"/>
</joint>

<link name="laser_frame_D1">
    <visual>
        <geometry>
            <box size="0.0002 0.0014 0.0026"/>
        </geometry>
        <material name="red"/>
    </visual>        
</link>



 <gazebo reference="laser_frame_D1">
    <material>Gazebo/Red</material>

    <sensor name="laser_D1" type="ray">
        <pose> 0 0 0 0 0 0 </pose>
        <visualize>false</visualize>
        <update_rate>100</update_rate>
        <ray>
            <scan>
                <horizontal>
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </horizontal>
                <vertical>
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </vertical>
            </scan>
            <range>
                <min>0.05</min>
                <max>2</max>
            </range>
        </ray>
        <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
            <ros>
                <argument>~/out:=scan</argument>
            </ros>
            <output_type>sensor_msgs/LaserScan</output_type>
            <topicName>laser_frame_D1</topicName>
        </plugin>
    </sensor>
</gazebo>

<joint name="laser_joint_D2" type="fixed">
    <parent link="wrist_3_link"/>
    <child link="laser_frame_D2"/>
    <origin xyz="-0.034 0.04 0.034" rpy="0.7458527 -0.2740142 1.8557205"/>
</joint> 

<link name="laser_frame_D2">
    <visual>
        <geometry>
            <box size="0.0002 0.0014 0.0026"/>
        </geometry>
        <material name="red"/>
    </visual>        
</link>

<gazebo reference="laser_frame_D2">
    <material>Gazebo/Red</material>

    <sensor name="laser_D2" type="ray">
        <pose> 0 0 0 0 0 0 </pose>
        <visualize>false</visualize>
        <update_rate>100</update_rate>
        <ray> <horizontal>
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </horizontal>
            <scan>
               
                <vertical>
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </vertical>
            </scan>
            <range>
                <min>0.05</min>
                <max>2</max>
            </range>
        </ray>
        <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
            <ros>
                <argument>~/out:=scan</argument>
            </ros>
            <output_type>sensor_msgs/LaserScan</output_type>
            <topicName>laser_frame_D2</topicName>
        </plugin>
    </sensor>
</gazebo> 
  <!-- <xacro:include filename="$(find myur5_description)/urdf/lidar/lidarC1.xacro" /> -->
  

  <!-- <xacro:include filename="$(find myur5_description)/urdf/lidar/lidarA2.xacro" /> -->
  <joint name="laser_joint_A2" type="fixed">
    <parent link="forearm_link"/>
    <child link="laser_frame_A2"/>
    <origin xyz="0.0 0.048 0.07" rpy="0 -1.1780972451 1.5707963268"/>
</joint>

<link name="laser_frame_A2">
    <visual>
        <geometry>
            <box size="0.0002 0.0026  0.0014"/>
        </geometry>
        <material name="red"/>
    </visual>        
</link>



<gazebo reference="laser_frame_A2">
    <material>Gazebo/Red</material>

    <sensor name="laser_A2" type="ray">
        <pose> 0 0 0 0 0 0 </pose>
        <visualize>false</visualize>
        <update_rate>100</update_rate>
        <ray>
            <scan>
                <horizontal>
                    <!--27 degrees-->
                    <samples>27</samples>
                    <min_angle>-0.1745329252  </min_angle> 
                    <max_angle>0.1745329252  </max_angle>
                </horizontal>
                <vertical>
                    <!--27 degrees-->
                    <samples>27</samples>
                    <min_angle>-0.1745329252  </min_angle> 
                    <max_angle>0.1745329252  </max_angle>
                </vertical>
            </scan>
            <range>
                <min>0.05</min>
                <max>2</max>
            </range>
        </ray>
        <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
            <ros>
                <argument>~/out:=scan</argument>
            </ros>
            <output_type>sensor_msgs/LaserScan</output_type>
            <topicName>laser_frame_A2</topicName>
        </plugin>
    </sensor>
</gazebo>
  <!-- <xacro:include filename="$(find myur5_description)/urdf/lidar/lidarB1.xacro" /> -->
  <joint name="laser_joint_B1" type="fixed">
    <parent link="wrist_3_link"/>
    <child link="laser_frame_B1"/>
    <origin xyz="0.034 0.04 0.034" rpy="-0.7458527 -0.2740142 1.2858722"/>
</joint>

<link name="laser_frame_B1">
    <visual>
        <geometry>
            <box size="0.0002 0.0014 0.0026"/>
        </geometry>
        <material name="red"/>
    </visual>        
</link>



<gazebo reference="laser_frame_B1">
    <material>Gazebo/Red</material>

    <sensor name="laser_B1" type="ray">
        <pose> 0 0 0 0 0 0 </pose>
        <visualize>false</visualize>
        <update_rate>100</update_rate>
        <ray>
            <scan>
                <horizontal>
                    <!--27 degrees-->
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </horizontal>
                <vertical>
                    <!--27 degrees-->
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </vertical>
            </scan>
            <range>
                <min>0.05</min>
                <max>2</max>
            </range>
        </ray>
        <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
            <ros>
                <argument>~/out:=scan</argument>
            </ros>
            <output_type>sensor_msgs/LaserScan</output_type>
            <topicName>laser_frame_B1</topicName>
        </plugin>
    </sensor>
</gazebo>
  <!-- <xacro:include filename="$(find myur5_description)/urdf/lidar/lidarB2.xacro" />  -->
<joint name="laser_joint_B2" type="fixed">
    <parent link="wrist_3_link"/>
    <child link="laser_frame_B2"/>
    <origin xyz="-0.034 0.04 -0.034" rpy="-0.7458527 0.2740142 1.8557205"/>
</joint>

<link name="laser_frame_B2">
    <visual>
        <geometry>
            <box size="0.0002 0.0014 0.0026"/>
        </geometry>
        <material name="red"/>
    </visual>        
</link>

<gazebo reference="laser_frame_B2">
    <material>Gazebo/Red</material>

    <sensor name="laser_B2" type="ray">
        <pose> 0 0 0 0 0 0 </pose>
        <visualize>false</visualize>
        <update_rate>100</update_rate>
        <ray>
            <scan>
                <horizontal>

                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </horizontal>
                <vertical>
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </vertical>
            </scan>
            <range>
                <min>0.05</min>
                <max>2</max>
            </range>
        </ray>
        <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
            <ros>
                <argument>~/out:=scan</argument>
            </ros>
            <output_type>sensor_msgs/LaserScan</output_type>
            <topicName>laser_frame_B2</topicName>
        </plugin>
    </sensor>
</gazebo>

<joint name="laser_joint_C2" type="fixed">
    <parent link="forearm_link"/>
    <child link="laser_frame_C2"/>
    <origin xyz="0.048 0.0 0.07" rpy="0 -1.9634954085 3.1415926536"/>
</joint>

<link name="laser_frame_C2">
    <visual>
        <geometry>
            <box size="0.0002 0.0026  0.0014"/>
        </geometry>
        <material name="red"/>
    </visual>        
</link>



<gazebo reference="laser_frame_C2">
    <material>Gazebo/Red</material>

    <sensor name="laser_C2" type="ray">
        <pose> 0 0 0 0 0 0 </pose>
        <visualize>false</visualize>
        <update_rate>100</update_rate>
        <ray>
            <scan>
                <horizontal>
                    <!--27 degrees-->
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </horizontal>
                <vertical>
                    <!--27 degrees-->
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </vertical>
            </scan>
            <range>
                <min>0.05</min>
                <max>2</max>
            </range>
        </ray>
        <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
            <ros>
                <argument>~/out:=scan</argument>
            </ros>
            <output_type>sensor_msgs/LaserScan</output_type>
            <topicName>laser_frame_C2</topicName>
        </plugin>
    </sensor>
</gazebo>

<joint name="laser_joint_C1" type="fixed">
  <parent link="forearm_link"/>
  <child link="laser_frame_C1"/>
  <origin xyz="-0.048 0.0 0.07" rpy="0 -1.1780972451 3.1415926536"/>
</joint>

<link name="laser_frame_C1">
  <visual>
      <geometry>
          <box size="0.0002 0.0026  0.0014"/>
      </geometry>
      <material name="red"/>
  </visual>        
</link>



<gazebo reference="laser_frame_C1">
  <material>Gazebo/Red</material>

  <sensor name="laser_C1" type="ray">
      <pose> 0 0 0 0 0 0 </pose>
      <visualize>true</visualize>
      <update_rate>100</update_rate>
      <ray>
          <scan>
              <horizontal>
                  <!--27 degrees-->
                  <samples>27</samples>
                  <min_angle>-0.1745329252 </min_angle>
                  <max_angle>0.1745329252 </max_angle>
              </horizontal>
              <vertical>
                  <!--27 degrees-->
                  <samples>27</samples>
                  <min_angle>-0.1745329252 </min_angle>
                  <max_angle>0.1745329252 </max_angle>
              </vertical>
          </scan>
          <range>
              <min>0.05</min>
              <max>2</max>
          </range>
      </ray>
      <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
          <ros>
              <argument>~/out:=scan</argument>
          </ros>
          <output_type>sensor_msgs/LaserScan</output_type>
          <topicName>laser_frame_C1</topicName>
      </plugin>
  </sensor>
</gazebo>

<joint name="laser_joint_A1" type="fixed">
    <parent link="forearm_link"/>
    <child link="laser_frame_A1"/>
    <origin xyz="0.0 -0.048 0.07" rpy="0 -1.9634954085 1.5707963268"/>
</joint>

<link name="laser_frame_A1">
    <visual>
        <geometry>
            <box size="0.0002 0.0026  0.0014"/>
        </geometry>
        <material name="red"/>
    </visual>        
</link>

<gazebo reference="laser_frame_A1">
    <material>Gazebo/Red</material>

     <sensor name="laser_A1" type="ray">
        <pose> 0 0 0 0 0 0 </pose>
        <visualize>false</visualize>
        <update_rate>100</update_rate>
        <ray>
            <scan>
                 <horizontal>
                    
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </horizontal> 
                <vertical>
                    
                    <samples>27</samples>
                    <min_angle>-0.1745329252 </min_angle> 
                    <max_angle>0.1745329252 </max_angle>
                </vertical>
            </scan>
            <range>
                <min>0.05</min>
                <max>2</max>
            </range>
        </ray>

        <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
            <ros>
                <argument>~/out:=scan</argument>
            </ros>
            <output_type>sensor_msgs/LaserScan</output_type>
            <topicName>laser_frame_A1</topicName>
        </plugin>
    </sensor>
</gazebo>

  <!-- test lidar -->
    <!-- <joint name="test_joint" type="fixed">
        <parent link="upper_arm_link"/>
        <child link="laser_frame_test"/>

        <origin xyz="0.055 0.0 0.24" rpy="0 0 0"/>
    </joint>

    <link name="laser_frame_test">
        <visual>
            <geometry>
                <box size="0.007 0.006  0.002"/>
            </geometry>
            <material name="red"/>
        </visual>       
    </link>



    <gazebo reference="laser_frame_test">
        <material>Gazebo/Red</material>

        <sensor name="laser_test" type="ray">
            <pose> 0 0 0 0 0 0 </pose>
            <visualize>false</visualize>
            <update_rate>100</update_rate>
            <ray>
                <scan>
                    <horizontal>
                    
                        <samples>27</samples>
                        <min_angle>-0.235619</min_angle> 
                        <max_angle>0.235619</max_angle>
                    </horizontal> 
                    <vertical>
                    
                        <samples>27</samples>
                        <min_angle>-0.235619</min_angle> 
                        <max_angle>0.235619</max_angle>
                    </vertical>
                </scan>
                <range>
                    <min>0.05</min>
                    <max>2</max>
                </range>
            </ray>

            <plugin name="laser_controller" filename="libgazebo_ros_laser.so">
                <ros>
                    <argument>~/out:=scan</argument>
                </ros>
                <output_type>sensor_msgs/LaserScan</output_type>
                <topicName>laser_frame_test</topicName>
            </plugin>
        </sensor>
    </gazebo> -->

  <!-- box -->
    <link name="box">
        <inertial>
            <!-- <origin xyz="-0.5 0.5 0.075" /> -->
            <origin xyz="0.01 0 -0.45" />
            <mass value="1.0" />
            <inertia  ixx="1.0" ixy="0.0"  ixz="0.0"  iyy="100.0"  iyz="0.0"  izz="1.0" />
        </inertial> 
        <visual>
            <!-- <origin xyz="-0.5 0.5 0.075" /> -->
            <origin xyz="0.01 0 -0.45" />
            <geometry>
                <!-- <box size="0.15 0.15 0.15"/> -->
                <mesh filename="package://ur5-joint-position-control/urdf/tekociTrak.STL" scale="1 1 1"/>
            </geometry>
            <material name="red"/>
        </visual>  
        <collision>
          <surface>
            <contact>
              <ode>
                <min_depth>1</min_depth>
                <max_vel>0.0</max_vel>
              </ode>
            </contact>
          </surface>
            <!-- <origin xyz="-0.5 0.5 0.075" /> -->
            <origin xyz="0.01 0 -0.45" />
            <geometry>
                <mesh filename="package://ur5-joint-position-control/urdf/tekociTrak.STL" scale="1 1 1"/>
                <!-- <box size="0.15 0.15 0.15"/> -->
            </geometry>
        </collision>   
    </link>
    <joint name="box_joint" type="fixed">
        <parent link="world"/>
        <child link="box"/>
        <!-- TODO: check base_link name of robot -->
        <origin rpy="1.5707963268 0.0 3.1415926536" xyz="0.0 0.0 0.0"/>
    </joint>

    <!-- define the ur5's position and orientation in the world coordinate system -->
    <link name="world"/>
    <joint name="world_joint" type="fixed">
        <parent link="world"/>
        <child link="base_link"/>
        <!-- TODO: check base_link name of robot -->
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
    </joint>
 

</robot>